src/neo/SmartContract/Native/NativeContract.cs

```diff
@@ -200,16 +200,26 @@ internal async void Invoke(ApplicationEngine engine, byte version)
                     throw new InvalidOperationException($"Cannot call this method with the flag {state.CallFlags}.");
                 engine.AddGas(method.CpuFee * engine.ExecFeeFactor + method.StorageFee * engine.StoragePrice);
                 List<object> parameters = new();
+                List<VM.Types.StackItem> native_stack = new();
                 if (method.NeedApplicationEngine) parameters.Add(engine);
                 if (method.NeedSnapshot) parameters.Add(engine.Snapshot);
                 for (int i = 0; i < method.Parameters.Length; i++)
-                    parameters.Add(engine.Convert(context.EvaluationStack.Pop(), method.Parameters[i]));
+                {
+                    var item = context.EvaluationStack.Pop();
+                    engine.ReferenceCounter.AddStackReference(item);
```

Comment:
IIUC after this change the items will be considered as referenced even though they're not on the stack and not in some slot. This is likely to work, but it doesn't look correct to me. To me it all should be symmetric, pushed items are added to the reference counter, popped items are removed from it. Native contracts consume parameters, they take them from the stack, so it's perfectly fine for the reference counter to go down to zero if there is nothing else left.

Reply:
When a native contract is called, it must hold the references to its parameters during method execution. It can be understood that the parameters are stored in a virtual `Arguments` slot during method execution.

Reply:
@roman-khimov Now I modified the logic. Please check it again.

Reply:
It's cleaner as an implementation, but I'd still rather pick the arguments from the stack.

And it looks to me that these items will be visible (as in present on the stack) to `onNEPXXPayment` handlers, they could change them.

Reply:
They can't. Because they are on the different context.